### PR TITLE
feat(protocol-designer): make timeline responsive

### PR DESCRIPTION
**Pull request by koji**
# Overview
make timeline responsive

close AUTH-877

## Changelog
- add DraggableSidebar component to make the sidebar draggable/resizable
- update AddStepButton to align with the design (remove + text and add + icon) 

## Risk assessment
low

### CODE REVIEW

File: protocol-designer/src/pages/Designer/ProtocolSteps/DraggableSidebar.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import {
+  DIRECTION_COLUMN,
+  DISPLAY_FLEX,
+  Flex,
+  JUSTIFY_SPACE_BETWEEN,
+  Box,
+} from '@opentrons/components'
+import { TimelineToolbox } from './Timeline/TimelineToolbox'
+
+const INITIAL_SIDEBAR_WIDTH = 350
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/DraggableSidebar.tsx]: i'll get the right value from Mel next week

File: protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx [hunk 2]
```diff
@@ -56,6 +59,7 @@ export function ProtocolSteps(): JSX.Element {
   const [deckView, setDeckView] = useState<
     typeof leftString | typeof rightString
   >(leftString)
+  const [targetWidth, setTargetWidth] = useState<number>(350)
```
Comment (koji) [inline protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx:62]: the initial value will be updated.

**Comment by koji** [comment]
- [x] fix check-js and lint-js errors
- [x] fix test errors
- [x] add test for DraggableSidebar

**Comment by ncdiehl11** [comment]
Whoa, looks really slick. I do see some highlighting when dragging the timeline toolbox (see "Ending deck"). Not sure if this is blocking, though.

https://github.com/user-attachments/assets/fd3ce6ce-7b32-48b7-9ceb-5a4832281194

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx [hunk 3]
```diff
@@ -183,14 +188,14 @@ export function StepContainer(props: StepContainerProps): JSX.Element {
 
   return (
     <>
-      {showDeleteConfirmation && (
+      {showDeleteConfirmation === true && (
```
Comment (jerader) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:192]: is this change needed?

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx [hunk 4]
```diff
@@ -183,14 +188,14 @@ export function StepContainer(props: StepContainerProps): JSX.Element {
...
         <ConfirmDeleteModal
           modalType={DELETE_STEP_FORM}
           onCancelClick={cancelDelete}
           onContinueClick={confirmDelete}
         />
       )}
-      {showMultiDeleteConfirmation && (
+      {showMultiDeleteConfirmation === true && (
```
Comment (jerader) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:199]: same

**Comment by ncdiehl11** [comment]
I think the add step menu is being contained by the timeline toolbox container
Screenshot 2024-12-17 at 10 21 02 AM

File: protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx [hunk 5]
```diff
@@ -74,6 +78,7 @@ export function ProtocolSteps(): JSX.Element {
     hasTimelineErrors && tab === 'protocolSteps' && formData == null
   const stepDetails = currentStep?.stepDetails ?? null
 
+  console.log('targetWidth', targetWidth)
```
Comment (jerader) [inline protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx]:
```suggestion
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx [hunk 6]
```diff
@@ -72,51 +74,11 @@ export function AddStepButton(): JSX.Element {
...
-    thermocycler: getIsModuleOnDeck(modules, THERMOCYCLER_MODULE_TYPE),
-    heaterShaker: getIsModuleOnDeck(modules, HEATERSHAKER_MODULE_TYPE),
-  }
-
   const addStep = (stepType: StepType): ReturnType<any> =>
     dispatch(stepsActions.addAndSelectStep({ stepType }))
 
-  const items = getSupportedSteps()
-    .filter(stepType => isStepTypeEnabled[stepType])
+  const items = getSupportedSteps(enableComment)
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx]: maybe
```suggestion
  getSupportedSteps(enableComment).
    reduce<JSX.Element[]>((acc, stepType) => {
      return getIsStepTypeEnabled(enableComment, modules)[stepType]
        ? [...acc, <AddOverflowButton ... />]
        : acc
    }, [])
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx [hunk 7]
```diff
@@ -179,6 +133,10 @@ export function AddStepButton(): JSX.Element {
         </Tooltip>
       )}
       <SecondaryButton
+        display="flex"
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx]:
```suggestion
        display={DISPLAY_FLEX}
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx [hunk 8]
```diff
@@ -187,8 +145,21 @@ export function AddStepButton(): JSX.Element {
...
 }
+
+const STEP_OVERFLOW_MENU_STYLE = css`
+  position: ${POSITION_ABSOLUTE};
+  z-index: 5;
+  left: 19.5rem;
+  white-space: ${NO_WRAP};
+  bottom: 4.2rem;
+  border-radius: ${BORDERS.borderRadius8};
+  box-shadow: 0px 1px 3px rgba(0, 0, 0, 0.2);
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx:209]: Do we have a constant for this?

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx [hunk 9]
```diff
@@ -48,9 +49,11 @@ import type { BaseState } from '../../../../types'
 const STARTING_DECK_STATE = 'Starting deck'
 const FINAL_DECK_STATE = 'Ending deck'
 const PX_HEIGHT_TO_TOP_OF_CONTAINER = 32
+const SIDEBAR_MIN_WIDTH_FOR_ICON = 179
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx]:
```suggestion
const PX_SIDEBAR_MIN_WIDTH_FOR_ICON = 179
```

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx [hunk 10]
```diff
@@ -21,6 +21,7 @@ import {
   toggleViewSubstep,
 } from '../../../../ui/steps/actions/actions'
 import { StepContainer } from './StepContainer'
+import { START_TERMINAL_ITEM_ID } from '../../../../steplist'
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx]: can move up a few imports

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx [hunk 11]
```diff
@@ -32,10 +33,12 @@ import { useTranslation } from 'react-i18next'
 
 export interface TerminalItemStepProps {
   id: TerminalItemId
+  title: string
```
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx]: is this used anywhere?

File: protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx (hunk 5 above)
Comment (ncdiehl11) [inline protocol-designer/src/pages/Designer/ProtocolSteps/index.tsx]: can remove

**Review by ncdiehl11 — CHANGES_REQUESTED** [review]
Left some comments but generally looking good

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx (hunk 3 above)
Comment (koji) [reply protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:192]: this is to solve eslint warning.

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx (hunk 4 above)
Comment (koji) [reply protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/StepContainer.tsx:199]: this is to solve eslint warning.

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx (hunk 8 above)
Comment (koji) [reply protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/AddStepButton.tsx:209]: you mean box-shadow or the entire style? 
i think we don't have the constant for this part.

File: protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx (hunk 11 above)
Comment (koji) [reply protocol-designer/src/pages/Designer/ProtocolSteps/Timeline/TerminalItemStep.tsx]: good catch. we don't need that anymore.

**Comment by koji** [comment]
> I think the add step menu is being contained by the timeline toolbox container Screenshot 2024-12-17 at 10 21 02 AM

@ncdiehl11 thank you for catching this.
finally, I could figure out what the issue is.

**Comment by ncdiehl11** [comment]
Screenshot 2024-12-20 at 2 56 49 PM
Do we want to break "Timeline"?

**Review by ncdiehl11 — COMMENTED** [review]
left one comment but looks good 💪

**Review by ncdiehl11 — APPROVED** [review]

**Comment by koji** [comment]
> Screenshot 2024-12-20 at 2 56 49 PM Do we want to break "Timeline"?

yeah, that is Mel's current design.
https://www.figma.com/design/WbkiUyU8VhtKz0JSuIFA45/Feature%3A-Protocol-Designer-Phase-1?node-id=10928-294030&m=dev

File: protocol-designer/src/pages/Designer/ProtocolSteps/DraggableSidebar.tsx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    }
+  }, [resize, stopResizing])
+
+  return (
+    <Flex
+      flexDirection={DIRECTION_COLUMN}
+      justifyContent={JUSTIFY_SPACE_BETWEEN}
+      height="100%"
+    >
+      <SidebarContainer ref={sidebarRef} resizedWidth={sidebarWidth}>
```
Comment (ddcc4) [inline protocol-designer/src/pages/Designer/ProtocolSteps/DraggableSidebar.tsx:70]: @koji Where is the `resizedWidth` property defined?

I can't seem to find any documentation about it, and it's causing the console to fill up with `Warning: React does not recognize the resizedWidth prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase resizedwidth instead. If you accidentally passed it from a parent component, remove it from the DOM element.`

**Pull request merged**